### PR TITLE
Added autograd support for C->C functions and enabled requires_grad=True for complex

### PR DESCRIPTION
**Pull request by anjali411**

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -3609,145 +3609,149 @@ def maybe_check_raise(fn, should_raise):
                  should_raise_tuple=(None, None, None))
 
     def test_autograd_simple_views_python(self):
```
Comment (albanD) [inline test/test_autograd.py:3611]: nit: to make the diff look better here, you can simply rename this to `_do_test_autograd_simple_views_python()` and then create two new tests that uses it.

File: torch/autograd/__init__.py [hunk 2]
```diff
@@ -28,6 +28,12 @@ def _make_grads(outputs, grads):
                                    + str(grad.shape) + " and output["
                                    + str(outputs.index(out)) + "] has a shape of "
                                    + str(out.shape) + ".")
+            if (out.dtype.is_complex and not grad.dtype.is_complex) or (not out.dtype.is_complex and grad.dtype.is_complex):
```
Comment (albanD) [inline torch/autograd/__init__.py:31]: I'm still not sure how this is different from `out.dtype.is_complex != grad.dtype.is_complex` ?

File: torch/autograd/__init__.py [hunk 3]
```diff
@@ -28,6 +28,12 @@ def _make_grads(outputs, grads):
...
+                raise RuntimeError("Mismatch in dtype: grad_output["
```
Comment (albanD) [inline torch/autograd/__init__.py:32]: nit: given that this is only for complex, we could be more precise here and replace `dtype` by stating that the user is mixing complex and real Tensors.

File: torch/autograd/gradcheck.py [hunk 4]
```diff
@@ -248,12 +248,12 @@ def fail_test(msg):
     for inp in tupled_inputs:
         if isinstance(inp, torch.Tensor):
             if inp.requires_grad:
-                if inp.dtype != torch.float64:
+                if inp.dtype != torch.float64 or inp.dtype != torch.complex128:
```
Comment (albanD) [inline torch/autograd/gradcheck.py:251]: This looks wrong no? It will always be True?
You want `not ( == or ==)` which translates to `(!= and !=)`.

File: torch/csrc/autograd/engine.cpp [hunk 5]
```diff
@@ -543,7 +543,12 @@ void validate_outputs(
       }
       grads[i] = at::sum_to(std::move(grads[i]), metadata.shape());
     }
-    TORCH_CHECK(isFloatingType(grads[i].scalar_type()));
+
+    bool input_is_complex = isComplexType(c10::typeMetaToScalarType(metadata.options().dtype()));
+    bool grad_is_complex = isComplexType(grads[i].scalar_type());
+
+    TORCH_CHECK(isFloatingType(grads[i].scalar_type()) || ((input_is_complex && grad_is_complex)
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp:550]: For the complex half, you simply want `input_is_complex == grad_is_complex` no?

File: torch/csrc/autograd/python_variable.cpp [hunk 6]
```diff
@@ -392,7 +392,7 @@ int THPVariable_set_requires_grad(THPVariable *self, PyObject *obj, void *unused
     THPUtils_setError(autograd::utils::requires_grad_leaf_error(obj == Py_True).c_str());
     return -1;
   }
-  if (requires_grad && !var.is_floating_point()) {
+  if (requires_grad && !(var.is_floating_point() || var.is_complex())) {
     THPUtils_setError("only Tensors of floating point dtype can require gradients");
```
Comment (albanD) [inline torch/csrc/autograd/python_variable.cpp:396]: nit: you might want to update this error message as well.

File: torch/csrc/autograd/variable.h [hunk 7]
```diff
@@ -213,7 +213,8 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
   /// variables.
   void set_requires_grad(bool requires_grad, at::TensorImpl* self_impl) override {
     TORCH_CHECK(
-      !requires_grad || at::isFloatingType(at::typeMetaToScalarType(self_impl->dtype())),
+      !requires_grad || at::isFloatingType(at::typeMetaToScalarType(self_impl->dtype())) ||
+      at::isComplexType(at::typeMetaToScalarType(self_impl->dtype())),
       "Only Tensors of floating point dtype can require gradients");
```
Comment (albanD) [inline torch/csrc/autograd/variable.h:218]: nit: same error message update

**Review by albanD — COMMENTED** [review]
This looks quite good for the enabling part.
I feel like we should add a big warning somewhere though that this is highly experimental and can go wrong in many ways.
Not sure if the `set_requires_grad()` is the right place? Or the `.backward()`? @gchanan  do you have an opinion on this?

File: torch/autograd/__init__.py (hunk 2 above)
Comment (anjali411) [reply torch/autograd/__init__.py:31]: updated

File: torch/csrc/autograd/engine.cpp (hunk 5 above)
Comment (anjali411) [reply torch/csrc/autograd/engine.cpp:550]: we currently don't support complex half

File: torch/autograd/gradcheck.py (hunk 4 above)
Comment (anjali411) [reply torch/autograd/gradcheck.py:251]: yeah you are right! thanks for the catch

File: torch/autograd/__init__.py (hunk 3 above)
Comment (gchanan) [reply torch/autograd/__init__.py:32]: I'm not sure I understand where this is used -- why is it okay for the dtypes to differ at all?

Comment (anjali411) [reply torch/autograd/__init__.py:32]: yeah I am not sure either. I saw some test failures when I enabled it more generally for all dtypes:
```
======================================================================
ERROR [0.003s]: test_no_grad (_main_.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_nn.py", line 629, in test_no_grad
    output.backward(torch.ones(1, 5, 10, 10))
  File "C:\Users\circleci\project\build\win_tmp\build\torch\tensor.py", line 198, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\autograd\__init__.py", line 100, in backward
    grad_tensors = _make_grads(tensors, grad_tensors)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\autograd\__init__.py", line 36, in _make_grads
    + str(out.dtype) + ".")
RuntimeError: Mismatch in dtype: grad_output[0] has a dtype of torch.float64 and output[0] has a dtype of torch.bfloat16.
```

Comment (anjali411) [reply torch/autograd/__init__.py:32]:
```
======================================================================
FAIL [0.002s]: test_vjp_err_check_strict (_main_.TestAutogradFunctional)
----------------------------------------------------------------------
RuntimeError: Mismatch in dtype: grad_output[0] has a dtype of torch.float64 and output[0] has a dtype of torch.float32.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_autograd.py", line 4352, in test_vjp_err_check_strict
    res = autogradF.vjp(bar, inp, v, strict=True)
AssertionError: "The output of the user-provided function is independent of input 0" does not match "Mismatch in dtype: grad_output[0] has a dtype of torch.float64 and output[0] has a dtype of torch.float32."
```

**Comment by gchanan** [comment]
> Not sure if the set_requires_grad() is the right place? Or the .backward()? @gchanan do you have an opinion on this?

I'd probably do it on backward; we want to eventually support it so the smallest blast radius place seems correct.

File: torch/csrc/autograd/engine.cpp (hunk 5 above)
Comment (anjali411) [reply torch/csrc/autograd/engine.cpp:550]: oh you meant the code related to complex and not the dtype complex half, I see haha. yup makes sense, updated!

File: torch/autograd/__init__.py (hunk 3 above)
Comment (albanD) [reply torch/autograd/__init__.py]: I guess for low precision training? forward could be done in float16 but backward in float32? Otherwise it prevent convergence?

File: test/test_autograd.py (hunk 1 above)
Comment (albanD) [reply test/test_autograd.py:3624]: My comment here was to re-use the original `test_autograd_simple_views_python` and rename it `_do_test_autograd_simple_views_python`. So that you don't have to re-indent the whole function. And so github won't see the whole function as changed.

File: torch/csrc/autograd/python_engine.cpp [hunk 8]
```diff
@@ -144,6 +144,10 @@ PyObject *THPEngine_run_backward(THPEngine *self, PyObject *args, PyObject *kwar
     THPUtils_assert(THPVariable_Check(_tensor), "element %d of tensors "
         "tuple is not a Tensor", i);
     auto& variable = ((THPVariable*)_tensor)->cdata;
+    if(variable.is_complex()) {
```
Comment (albanD) [inline torch/csrc/autograd/python_engine.cpp:147]: Given that the code above allow complex backward for both the python and the cpp api. Maybe you want to add the same thing to the regular engine run_backward function.

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 9]
```diff
@@ -86,6 +86,9 @@ static inline Variable valueToTensor(c10::TensorOptions options, PyObject* value
   if (PyFloat_Check(value)) {
     return at::indexing::scalarToTensor(Scalar(THPUtils_unpackDouble(value)), options, device);
   }
+  if (PyComplex_Check(value)) {
```
Comment (albanD) [inline torch/csrc/autograd/python_variable_indexing.cpp:89]: Why is this needed now and was not before?

**Review by albanD — COMMENTED** [review]
Looks good. Just minor questions.

File: test/test_autograd.py (hunk 1 above)
Comment (anjali411) [reply test/test_autograd.py:3624]: oh okay updated!

File: torch/csrc/autograd/python_variable_indexing.cpp (hunk 9 above)
Comment (anjali411) [reply torch/csrc/autograd/python_variable_indexing.cpp:89]: this is used in the assignment of complex scalar to a variable in `get_numerical_jacobian` method [here](https://github.com/pytorch/pytorch/blob/3799d1d74a39cd3d4419539453acc7f5f7c63f7e/torch/autograd/gradcheck.py#L131-L136)

**Review by albanD — APPROVED** [review]
Looks good !
